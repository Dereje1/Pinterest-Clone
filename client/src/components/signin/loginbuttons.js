import React from 'react';
import PropTypes from 'prop-types';
import Button from '@mui/material/Button';
import TwitterIcon from '@mui/icons-material/Twitter';
import GoogleIcon from '@mui/icons-material/Google';
import GitHubIcon from '@mui/icons-material/GitHub';
import AccountCircleIcon from '@mui/icons-material/AccountCircle';
import { connect } from 'react-redux';
import { setGuest } from '../../actions/authentication';
import './loginbuttons.scss';

export const mapStateToProps = ({ user }) => ({ user });

const actionCreators = {
  setGuest,
};

const handleLogin = (loc) => { // twitter/ google authentication
  window.location.assign(loc);
};

export const ProviderButton = ({ service }) => {
  const providerIcons = {
    twitter: <TwitterIcon style={{ fontSize: 25 }} />,
    google: <GoogleIcon style={{ fontSize: 25 }} />,
    github: <GitHubIcon style={{ fontSize: 25 }} />,
  };
  return (
    <Button
      id={`${service}loginbutton`}
      variant="outlined"
      startIcon={providerIcons[service]}
      onClick={() => handleLogin(`/auth/${service}`)}
      onMouseDown={e => e.preventDefault()}
    >
      {`Continue With ${service.charAt(0).toUpperCase() + service.slice(1)}`}
    </Button>
  );
};

export class LoginButtons extends React.Component {

  handleGuest = () => { // set guest user
    const { setGuest: setGuestStatus, guest } = this.props;
    setGuestStatus('/auth/guest');
    guest(); // callback to hid login div
  };

  render() {
    const { user: { providers, username } } = this.props;
    if (!providers) return null;
    const providerKeys = Object.keys(providers);
    return (
      <React.Fragment>
        {

          username !== 'Guest' && (
            <Button id="guestbutton" variant="outlined" startIcon={<AccountCircleIcon style={{ fontSize: 25 }} />} onClick={this.handleGuest}>
              Continue As Guest&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
            </Button>
          )
        }

        {
          providerKeys.map((service) => {
            if (providers[service]) return <ProviderButton key={service} service={service} />;
            return null;
          })
        }
      </React.Fragment>
    );
  }

}

export default connect(mapStateToProps, actionCreators)(LoginButtons);

LoginButtons.defaultProps = {
  user: {},
};

LoginButtons.propTypes = {
  // redux action to set guest user status on server
  setGuest: PropTypes.func.isRequired,
  // used for reseting back to guest mode in signin
  guest: PropTypes.func.isRequired,
  user: PropTypes.shape(PropTypes.shape),
};

ProviderButton.propTypes = {
  service: PropTypes.string.isRequired,
};
