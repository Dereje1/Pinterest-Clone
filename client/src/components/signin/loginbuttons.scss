#twitterloginbutton, #googleloginbutton,#guestbutton, #githubloginbutton{
    border: 0px;
    height: 40px;
    display: flex;
    justify-content: space-around;
    align-items: center;
    border-radius: 2px;
    padding: 1px;
    font-size: 15px;
    font-weight: bold;
    outline: none;
    box-shadow: none;
    cursor: pointer;
    background-clip: padding-box;
    transition: opacity .2s linear;
    width: 268px;
    margin-bottom: 10px;
    color:white;
    #twitter{
      font-size: 30px;
      padding-left: 3px;
    }
    #guest{
        font-size: 30px;
        padding-left: 3px;
      }
    .buttontext{
        padding-right: 45px;
      }
  }
  #googleloginbutton{
    background-color: #4285F4;
  }
  #twitterloginbutton{
    background-color: #1DA1F2;
  }
  #guestbutton{
    background-color: #ffb45e;
  }
  #githubloginbutton{
    background-color: #1d7b20;
  }