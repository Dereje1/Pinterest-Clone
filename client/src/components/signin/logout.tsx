import { useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import { useDispatch } from 'react-redux';
import { getUser } from '../../actions/authentication';
import RESTcall from '../../crud';

function LogOut() {
  const history = useHistory();
  const dispatch = useDispatch();

  const logout = async () => {
    // TODO: vite erroring out on axios rest call to auth/logout
    // await RESTcall({ address: '/auth/logout' });
    // dispatch(getUser('/auth/profile'));
    // history.replace('/logout');
    window.location.assign('/auth/logout');
  };

  useEffect(() => {
    logout();
  }, []);

  return null;
}

export default LogOut;
