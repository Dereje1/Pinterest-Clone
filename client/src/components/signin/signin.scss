@import "../../keyframes.scss";

.signinmodal{
    position: fixed;
    left: 0;
    right: 0;
    width: 370px;
    height: 250px;
    z-index: 1500;
    margin-left: auto;
    margin-right: auto;
    padding-top: 10px;
    background-color: #0b0f19c4;
    border-radius: 8px;
    display: flex;
    flex-direction: column;
    justify-content: space-around;
    align-items: center;
    border-radius: 10px;
    &.signshow{
        @include animation('dropdown');
        animation-duration: 1s;
        animation-timing-function: cubic-bezier();
        animation-iteration-count: 1;
        top: 200px;
        visibility: visible;
    }
    &.signhide{
        @include animation('pullup');
        animation-duration: 1s;
        animation-timing-function: linear;
        animation-iteration-count: 1;
    }
}

@include keyframes(dropdown){
    0%{
        transform: translateY(-300px);
    }
    100%{
        transform: translateY(0px);
    }
}

@include keyframes(pullup){
    0%{
        transform: translateY(200px);
        visibility: visible;
    }
    100%{
        transform: translateY(-300px);
    }
}