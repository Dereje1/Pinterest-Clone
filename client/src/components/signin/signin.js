import React from 'react';
import PropTypes from 'prop-types';
import LoginButtons from './loginbuttons';
import {
  delay,
} from '../../utils/utils';
import './signin.scss';

export class SignIn extends React.Component {

  constructor(props) {
    super(props);
    this.state = {
      show: false,
    };
    this.signInModal = React.createRef();
  }

  componentDidMount() {
    this.signInModal.current.focus();
    this.setState({ show: true });
  }

  resetGuest = () => {
    const { removeSignin } = this.props;
    this.setState({
      show: false,
    }, async () => {
      await delay(1000);
      removeSignin();
    });
  };

  render() {
    const { show } = this.state;
    return (
      <>
        {show && <div className="modal-overlay" />}
        <div
          className={show ? 'signin signshow' : 'signin signhide'}
          ref={this.signInModal}
          onBlur={this.resetGuest}
          tabIndex={0}
          role="menuitem"
        >
          <LoginButtons guest={this.resetGuest} />
        </div>
      </>
    );
  }

}
export default SignIn;

SignIn.propTypes = {
  // callback to toggle login div
  removeSignin: PropTypes.func.isRequired,
  // used for eventlistener targeting to close login div
};
