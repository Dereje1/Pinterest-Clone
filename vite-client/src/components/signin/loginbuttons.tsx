import React from 'react';
import Button from '@mui/material/Button';
import AccountCircleIcon from '@mui/icons-material/AccountCircle';
import { connect } from 'react-redux';
import { setGuest } from '../../actions/authentication';
import { getProviderIcons } from '../common/common';
import { userType, providerIconsType, providersType } from '../../interfaces';
import './loginbuttons.scss';

export const mapStateToProps = ({ user }: {user: userType}) => ({ user });

const actionCreators = {
  setGuest,
};

const handleLogin = (loc: string) => { // twitter/ google authentication
  window.location.assign(loc);
};

export function ProviderButton({ service }: {service: string}) {
  const providerIcons: providerIconsType = getProviderIcons({ fontSize: 25 });
  return (
    <Button
      id={`${service}loginbutton`}
      variant="outlined"
      startIcon={providerIcons[service as keyof providerIconsType].icon}
      onClick={() => handleLogin(`/auth/${service}`)}
      onMouseDown={(e) => e.preventDefault()}
    >
      {`Continue With ${service.charAt(0).toUpperCase() + service.slice(1)}`}
    </Button>
  );
}

interface LoginButtonsProps{
  user: userType
  setGuest: (path: string) => void
  guest: () => void
}

export class LoginButtons extends React.Component<LoginButtonsProps> {

  handleGuest = () => { // set guest user
    const { setGuest: setGuestStatus, guest } = this.props;
    setGuestStatus('/auth/guest');
    guest(); // callback to hid login div
  };

  render() {
    const { user: { providers, username } } = this.props;
    if (!providers) return null;
    const providerKeys = Object.keys(providers);
    return (
      <>
        {

          username !== 'Guest' && (
            <Button id="guestbutton" variant="outlined" startIcon={<AccountCircleIcon style={{ fontSize: 25 }} />} onClick={this.handleGuest}>
              Continue As Guest&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
            </Button>
          )
        }

        {
          providerKeys.map((service) => {
            if (providers[service as keyof providersType]) {
              return <ProviderButton key={service} service={service} />;
            }
            return null;
          })
        }
      </>
    );
  }

}

export default connect(mapStateToProps, actionCreators)(LoginButtons);
