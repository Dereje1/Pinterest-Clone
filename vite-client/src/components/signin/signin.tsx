import React, { createRef } from 'react';
import LoginButtons from './loginbuttons';
import {
  delay,
} from '../../utils/utils';
import './signin.scss';

interface SignInProps{
  removeSignin: () => void
}

interface SignInState {
  show: boolean
}

export class SignIn extends React.Component<SignInProps, SignInState> {

  private signInModal = createRef<HTMLDivElement>();

  constructor(props: SignInProps) {
    super(props);
    this.state = {
      show: false,
    };
  }

  componentDidMount() {
    if (this.signInModal.current) {
      this.signInModal.current.focus();
    }
    this.setState({ show: true });
  }

  resetGuest = () => {
    const { removeSignin } = this.props;
    this.setState({
      show: false,
    }, async () => {
      await delay(1000);
      removeSignin();
    });
  };

  render() {
    const { show } = this.state;
    return (
      <>
        {show && <div className="modal-overlay" />}
        <div
          className={show ? 'signinmodal signshow' : 'signinmodal signhide'}
          ref={this.signInModal}
          onBlur={this.resetGuest}
          tabIndex={0}
          role="menuitem"
        >
          <LoginButtons guest={this.resetGuest} />
        </div>
      </>
    );
  }

}
export default SignIn;
