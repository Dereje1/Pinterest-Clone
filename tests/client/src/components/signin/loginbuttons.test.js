/**
 * @jest-environment jsdom
 */
import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { LoginButtons } from '../../../../../client/src/components/signin/loginbuttons';

describe('The sign in component', () => {
  let props;
  let mockedAssign;
  beforeEach(() => {
    props = {
      setGuest: jest.fn(),
      guest: jest.fn(),
    };
    const windowSpy = jest.spyOn(global, 'window', 'get');
    mockedAssign = jest.fn();
    windowSpy.mockImplementation(() => ({
      location: {
        assign: mockedAssign,
      },
    }));
  });

  afterEach(() => {
    props = null;
    jest.clearAllMocks();
  });

  test('will render', () => {
    const wrapper = shallow(<LoginButtons {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  test('will handle guest users', () => {
    const wrapper = shallow(<LoginButtons {...props} />);
    const guestButton = wrapper.find({ id: 'guestbutton' });
    guestButton.props().onClick();
    expect(props.setGuest).toHaveBeenCalledWith('/auth/guest');
    expect(props.guest).toHaveBeenCalledTimes(1);
  });

  test('will handle twitter users', () => {
    const wrapper = shallow(<LoginButtons {...props} />);
    const twitterButton = wrapper.find({ id: 'twitterloginbutton' });
    twitterButton.props().onClick();
    expect(mockedAssign).toHaveBeenCalledTimes(1);
    expect(mockedAssign).toHaveBeenCalledWith('/auth/twitter');
  });

  test('will handle google users', () => {
    const wrapper = shallow(<LoginButtons {...props} />);
    const googleButton = wrapper.find({ id: 'googleloginbutton' });
    googleButton.props().onClick();
    expect(mockedAssign).toHaveBeenCalledTimes(1);
    expect(mockedAssign).toHaveBeenCalledWith('/auth/google');
  });

  test('will handle github users', () => {
    const wrapper = shallow(<LoginButtons {...props} />);
    const googleButton = wrapper.find({ id: 'githubloginbutton' });
    googleButton.props().onClick();
    expect(mockedAssign).toHaveBeenCalledTimes(1);
    expect(mockedAssign).toHaveBeenCalledWith('/auth/github');
  });
});
