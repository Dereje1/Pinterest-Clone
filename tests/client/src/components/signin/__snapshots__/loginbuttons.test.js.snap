// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The sign in component will render for Guest users on homepage 1`] = `
<Fragment>
  <ProviderButton
    key="twitter"
    service="twitter"
  />
  <ProviderButton
    key="google"
    service="google"
  />
  <ProviderButton
    key="github"
    service="github"
  />
</Fragment>
`;

exports[`The sign in component will render for cover page 1`] = `
<Fragment>
  <ForwardRef(Button)
    id="guestbutton"
    onClick={[Function]}
    startIcon={
      <Memo
        style={
          Object {
            "fontSize": 25,
          }
        }
      />
    }
    variant="outlined"
  >
    Continue As Guest        
  </ForwardRef(Button)>
  <ProviderButton
    key="twitter"
    service="twitter"
  />
  <ProviderButton
    key="google"
    service="google"
  />
  <ProviderButton
    key="github"
    service="github"
  />
</Fragment>
`;

exports[`The sign in component will render only available providers 1`] = `
<Fragment>
  <ForwardRef(Button)
    id="guestbutton"
    onClick={[Function]}
    startIcon={
      <Memo
        style={
          Object {
            "fontSize": 25,
          }
        }
      />
    }
    variant="outlined"
  >
    Continue As Guest        
  </ForwardRef(Button)>
  <ProviderButton
    key="github"
    service="github"
  />
</Fragment>
`;
